pulp_rpm/app/modulemd.py

```diff
@@ -18,6 +18,12 @@ def resolve_module_packages(version, previous_version):
     Args:
         version (pulpcore.app.models.RepositoryVersion): current incomplete repository version
     """
+    def modules_packages(modules):
+        packages = set()
+        for module in modules:
```

Comment:
@daviddavis wdyt? i thin this way we'll have less qs evaluated?

```
$ git diff
diff --git a/pulp_rpm/app/modulemd.py b/pulp_rpm/app/modulemd.py
index fc999434..83dbdf46 100644
--- a/pulp_rpm/app/modulemd.py
+++ b/pulp_rpm/app/modulemd.py
@@ -3,7 +3,7 @@ import os
 import tempfile
 
 from pulpcore.app.models.content import Artifact
-from pulp_rpm.app.models import Modulemd
+from pulp_rpm.app.models import Modulemd, Package
 from pulp_rpm.app.constants import PULP_MODULE_ATTR, PULP_MODULEDEFAULTS_ATTR
 
 import gi
@@ -21,22 +21,26 @@ def resolve_module_packages(version, previous_version):
     current_modules = Modulemd.objects \
         .filter(pk__in=version.content.filter(pulp_type="rpm.modulemd"))
     if previous_version:
         previous_modules = Modulemd.objects \
             .filter(pk__in=previous_version.content.filter(pulp_type="rpm.modulemd"))
         added_modules = current_modules.difference(previous_modules)
         removed_modules = previous_modules.difference(current_modules)
-        current_module_packages = current_modules.values_list("packages", flat=True)
-        removed_module_packages = removed_modules.values_list("packages", flat=True)
-        packages_to_remove = removed_module_packages.difference(current_module_packages)
-        version.remove_content(packages_to_remove)
+        current_module_packages = current_modules.values_list("packages", flat=True).distinct()
+        removed_module_packages = removed_modules.values_list("packages", flat=True).distinct()
+        packages_to_remove = set(removed_module_packages.difference(current_module_packages))
+        packages_to_remove.discard(None)
+        version.remove_content(Package.objects.filter(pk__in=packages_to_remove))
     else:
         added_modules = current_modules
 
-    added_module_packages = added_modules.values_list("packages", flat=True)
-    current_packages = version.content.filter(pulp_type="rpm.package")
-    packages_to_add = added_module_packages.difference(current_packages)
-    version.add_content(packages_to_add)
+    added_module_packages = added_modules.values_list("packages", flat=True).distinct()
+    current_packages = version.content.filter(pulp_type="rpm.package").values_list('pk')
+    packages_to_add = set(added_module_packages.difference(current_packages))
+    packages_to_add.discard(None)
+    version.add_content(Package.objects.filter(pk__in=packages_to_add))
```

Reply:
@ipanova the problem is not that the units aren't unique. Rather, it's that it's not possible to use `values_list()` on a many-to-many. [The django docs for values_list() call this out](https://docs.djangoproject.com/en/2.2/ref/models/querysets/#values-list) but here's also an example. Notice that kangaroo isn't being returned as part of the values_list.

```
In [35]: modules = Modulemd.objects.all()[5:7]

In [36]: modules
Out[36]: <QuerySet [<Modulemd: duck>, <Modulemd: kangaroo>]>

In [37]: modules[0].packages.all()
Out[37]: <QuerySet [<Package: duck>, <Package: frog>]>

In [38]: modules[1].packages.all()
Out[38]: <QuerySet [<Package: kangaroo>]>

In [39]: modules.values_list("packages", flat=True)
Out[39]: <QuerySet [UUID('dbe19adc-2a68-4a3c-9a9a-115add12d656'), UUID('83bd86f0-ec2d-49c1-9ece-2e07f111bcb4')]>

In [40]: modules.values_list("packages")
Out[40]: <QuerySet [(UUID('dbe19adc-2a68-4a3c-9a9a-115add12d656'),), (UUID('83bd86f0-ec2d-49c1-9ece-2e07f111bcb4'),)]>

In [41]: Package.objects.filter(pk__in=modules.values_list("packages", flat=True))
Out[41]: <QuerySet [<Package: duck>, <Package: frog>]>
```